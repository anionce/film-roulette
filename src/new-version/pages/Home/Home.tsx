import React, { ChangeEvent, useState } from 'react';
import './Home.scss';

import { ButtonGroup } from '../../components/ButtonGroup/ButtonGroup';
import { FilterType, filterTypes } from '../../../constants/filters';
import { PlayButton } from '../../components/PlayButton/PlayButton';
import { FiltersModal } from '../../components/FiltersModal/FiltersModal';
import { MovieGenre } from '../../../constants/genre';
import { MovieRuntime } from '../../../constants/runtime';
import { StreamingServices } from '../../../constants/streamingServices';

export type FilterArguments = {
	genre: MovieGenre | null;
	duration: MovieRuntime | null;
	streaming: StreamingServices[] | null;
};

type ModalOpen = {
	[FilterType.Genre]: boolean;
	[FilterType.Duration]: boolean;
	[FilterType.Streaming]: boolean;
};

export type HomeProps = {
	filters: FilterArguments;
	onButtonClick: any;
	setFilters: any;
};

export const Home = ({ filters, onButtonClick, setFilters }: HomeProps) => {
	const [open, setOpen] = useState<ModalOpen>({
		[FilterType.Genre]: false,
		[FilterType.Duration]: false,
		[FilterType.Streaming]: false,
	});

	const openModal = (type: FilterType): void => {
		setOpen(prev => ({ ...prev, [type]: true }));
	};

	const closeModal = (type: FilterType) => {
		setOpen(prev => ({ ...prev, [type]: false }));
	};

	const onSelectGenre = (event: React.MouseEvent<HTMLButtonElement>) => {
		onGenreChange(event);
		setTimeout(() => closeModal(FilterType.Genre), 250);
	};

	const onGenreChange = (event: ChangeEvent<HTMLInputElement> | React.MouseEvent<HTMLButtonElement>): void => {
		const value = event.currentTarget.getAttribute('data-value');
		setFilters((prev: FilterArguments) => ({ ...prev, genre: value as MovieGenre }));
	};

	const getSelect = (type: FilterType) => {
		const options = {
			[FilterType.Genre]: onSelectGenre,
			[FilterType.Duration]: onSelectDuration,
			[FilterType.Streaming]: selectedServicesOnChange,
		};

		return options[type];
	};

	const onDurationChange = (event: ChangeEvent<HTMLInputElement> | React.MouseEvent<HTMLButtonElement>): void => {
		const value = event.currentTarget.getAttribute('data-value');
		setFilters((prev: FilterArguments) => ({ ...prev, duration: value as MovieRuntime }));
	};

	const onSelectDuration = (event: React.MouseEvent<HTMLButtonElement>) => {
		onDurationChange(event);
		setTimeout(() => closeModal(FilterType.Duration), 250);
	};

	const selectedServicesOnChange = (event: ChangeEvent<HTMLInputElement>, newServices: StreamingServices[]): void => {
		setFilters((prev: FilterArguments) => ({ ...prev, streaming: newServices }));
	};

	const isDisabled = !filters.streaming?.length;

	return (
		<div className='home-mobile-container'>
			<h2 className='home-mobile-title'>¡Hola!</h2>
			<div className='home-mobile-subtitle'>
				Elige qué película ver según el género que te apetezca y el tiempo que tengas &#129303;
			</div>
			<ButtonGroup filters={filters} openModal={openModal} />
			<PlayButton filters={filters} onButtonClick={onButtonClick} />
			{filterTypes.map(type => {
				return (
					<FiltersModal
						key={`filter-modal-${type}`}
						open={open[type]}
						filterType={type}
						closeModal={closeModal}
						onSelect={getSelect(type)}
						filters={filters.streaming}
						isButtonDisabled={isDisabled}
					/>
				);
			})}
		</div>
	);
};
