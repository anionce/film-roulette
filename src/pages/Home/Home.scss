@import '../../assets/styles/variables';

.home-container {
	display: flex;
	height: 100vh;
	width: 100%;
	font-weight: 700;
}

.home-left-side {
	width: 50%;
	display: flex;
	gap: 50px;
	flex-direction: column;
	justify-content: center;
	align-items: center;
	background-color: $color-background-gray;
}

.home-left-container {
	display: flex;
	gap: 50px;
	flex-direction: column;
}

.home-right-side {
	width: 50%;
	display: flex;
	justify-content: center;
	align-items: center;
	background-color: $color-yellow;
}

.home-right-container {
	display: flex;
	flex-direction: column;
	align-items: center;
}

.home-subtitle {
	text-transform: uppercase;
	color: $color-yellow;
	margin: 0;
	font-size: $font-size-sm;
}

.home-label {
	font-family: $main-font;
	font-size: $font-size-xxl;
	margin: 0;
}

.home-label-middle {
	text-align: end;
}

.error-input {
	border: 1px red solid;
}

option {
	width: 500px;
	height: 60px;
	padding: 8px;
	font-size: $font-size-md;
	font-family: $main-font;
	font-weight: 700;
}

button {
	background-color: $color-yellow;
	border: 1px solid black;
	width: 150px;
	padding: 20px;
	font-family: $main-font;
	font-weight: 700;
	font-size: $font-size-lg;
	cursor: pointer;
	align-self: center;
	border-radius: 4px;
}

.additional-info {
	display: flex;
	flex-direction: column;
	align-items: center;
}

@media only screen and (max-width: 1100px) {
	.home-container {
		display: flex;
		flex-direction: column;
	}

	.home-left-container {
		gap: 8px;
	}

	.home-left-side {
		width: 100%;
		height: 50%;
	}

	.home-section-top,
	.home-section-middle {
		display: flex;
		flex-direction: column;
		align-items: center;
		gap: 8px;
	}

	.home-right-side {
		width: 100%;
		height: 50%;
		display: 'flex';
	}

	.home-right-container {
		padding: 16px;
		display: flex;
		flex-direction: row;
		justify-content: center;
	}

	.home-subtitle {
		font-size: $font-size-xxl;
		margin-bottom: 16px;
	}

	.home-label {
		font-size: $font-size-xl;
		text-align: center;
	}

	.additional-info {
		gap: 16px;
	}

	button {
		margin-top: 16px;
		display: flex;
		align-items: center;
		width: 80px;
		height: 30px;
		font-size: $font-size-sm;
	}
}

@media only screen and (max-width: 500px) {
	.home-right-container {
		display: flex;
		flex-direction: column;
		align-items: center;
	}

	.home-right-side {
		width: 100%;
		height: 80%;
		display: 'flex';
	}

	.home-label {
		font-size: $font-size-lg;
		text-align: center;
	}
	.additional-info {
		gap: 0;
	}
}
