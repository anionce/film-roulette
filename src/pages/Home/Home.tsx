import React, { useEffect, useState } from 'react';
import { Movie } from '../../models/MovieResponse';
import { useLazyGetDetailsQuery, useLazyGetRandomMovieQuery } from '../../services/api';
import { Loader } from '../../components/Loader/Loader';
import { Poster } from '../../components/Poster/Poster';
import { MovieGenre, mapValueToGenre } from '../../constants/genre';
import { MovieRuntime, mapValueToMovieRuntime } from '../../constants/runtime';
import { GenreSelection } from '../../components/GenreSelection/GenreSelection';
import { RuntimeSelection } from '../../components/RuntimeSelection/RuntimeSelection';
import './Home.scss';
import { Plot } from '../../components/Plot/Plot';
import { Rating } from '../../components/Rating/Rating';

export const Home = () => {
	const [randomMovie, setRandomMovie] = useState<Movie | undefined>(undefined);

	const [duration, setDuration] = useState<MovieRuntime | undefined>(MovieRuntime.Short);
	const [genre, setGenre] = useState<MovieGenre | undefined>(MovieGenre.Action);
	const [actualPage, setActualPage] = useState<number>(1);

	const [trigger, { data: dataRandom, isLoading }] = useLazyGetRandomMovieQuery();
	const [triggerIMDBDetail, { data: dataIMDB }] = useLazyGetDetailsQuery();

	const shouldShowPoster: boolean = !!randomMovie && !!dataIMDB;

	useEffect(() => {
		if (dataRandom) {
			const { total_pages, results } = dataRandom;
			setActualPage(prevPage => (total_pages === actualPage ? 1 : prevPage + 1));
			setRandomMovie(results[Math.floor(Math.random() * results.length)]);
		}
		// eslint-disable-next-line
	}, [dataRandom]);

	useEffect(() => {
		if (randomMovie) {
			triggerIMDBDetail({ id: randomMovie.id });
		}
		// eslint-disable-next-line
	}, [randomMovie]);

	const onButtonClick = () => {
		if (duration && genre) {
			trigger({
				page: actualPage,
				runtime: mapValueToMovieRuntime(duration),
				genres: mapValueToGenre(genre),
			});
		}
	};

	const onDurationChange = (event: React.ChangeEvent<HTMLSelectElement>): void => {
		setActualPage(1);
		setDuration(event.target.value as MovieRuntime);
	};

	const onGenreChange = (event: React.ChangeEvent<HTMLSelectElement>): void => {
		setActualPage(1);
		setGenre(event.target.value as MovieGenre);
	};

	return (
		<div className='home-container'>
			<div className='home-left-side'>
				<div className='home-left-container'>
					<GenreSelection onMainGenreChange={onGenreChange} />
					<RuntimeSelection onDurationChange={onDurationChange} />
					<button onClick={onButtonClick}>PLAY</button>
				</div>
			</div>
			<div className='home-right-side'>
				{isLoading && <Loader />}
				{shouldShowPoster && (
					<div className='home-right-container'>
						<Poster dataIMDB={dataIMDB as string} randomMovie={randomMovie as Movie} />
						<div className='additional-info'>
							<Plot plot={(randomMovie as Movie)?.overview} dataIMDB={dataIMDB as string} />
							<Rating rating={(randomMovie as Movie)?.vote_average} />
						</div>
					</div>
				)}
			</div>
		</div>
	);
};
