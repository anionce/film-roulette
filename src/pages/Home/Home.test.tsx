import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import { customRender } from '../../assets/tests/testUtils';
import { Home } from './Home';

test('renders correcly', () => {
	const { container } = customRender(<Home />);

	expect(container).toMatchSnapshot();
});
