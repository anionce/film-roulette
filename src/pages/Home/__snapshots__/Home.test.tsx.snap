// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correcly 1`] = `
<div>
  <div
    class="home-container"
  >
    <div
      class="home-left-side"
    >
      <div
        class="home-left-container"
      >
        <div
          class="home-section-top"
        >
          <p
            class="home-label"
          >
            Hoy me apetece...
          </p>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary root-element css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
          >
            <div
              aria-controls=":r0:"
              aria-expanded="false"
              aria-haspopup="listbox"
              class="MuiSelect-select select-element MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
              role="combobox"
              tabindex="0"
            >
              <span
                class="notranslate"
              >
                ​
              </span>
            </div>
            <input
              aria-hidden="true"
              aria-invalid="false"
              class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
              tabindex="-1"
              value=""
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-ihdtdm"
              >
                <span
                  class="notranslate"
                >
                  ​
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="home-section-middle"
        >
          <p
            class="home-label home-label-middle"
          >
            y tengo...
          </p>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary root-element css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
          >
            <div
              aria-controls=":r1:"
              aria-expanded="false"
              aria-haspopup="listbox"
              class="MuiSelect-select select-element MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
              role="combobox"
              tabindex="0"
            >
              <span
                class="notranslate"
              >
                ​
              </span>
            </div>
            <input
              aria-hidden="true"
              aria-invalid="false"
              class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
              tabindex="-1"
              value=""
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-ihdtdm"
              >
                <span
                  class="notranslate"
                >
                  ​
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <button>
          PLAY
        </button>
      </div>
    </div>
    <div
      class="home-right-side"
    />
  </div>
</div>
`;
